File: server/internal/handler/file.go
package handler

import (
	"net/http"
	"strings"
	"time"

	"server/internal/model"
	"server/internal/repository"
	"server/pkg/api"

	"github.com/gin-gonic/gin"
)

// GetFileContent GET /api/files/*fileId  (path ends with /content)
func GetFileContent(c *gin.Context) {
	raw := strings.TrimPrefix(c.Param("fileId"), "/")
	// 支持两种格式：
	// - /api/files/note.md/content  → fileId = "note.md"
	// - /api/files/h1/note.md/content → fileId = "h1/note.md"
	fileID := strings.TrimSuffix(raw, "/content")
	if fileID == "" {
		api.BadRequest(c, "fileId is required")
		return
	}

	repo := repository.GetStorage()
	content, ok := repo.GetFileContent(fileID)
	if !ok {
		api.NotFound(c, "file not found")
		return
	}
	api.Success(c, model.GetFileContentResponse{ID: fileID, Content: content})
}

// SaveFileContent PUT /api/files/*fileId  (path ends with /content)
func SaveFileContent(c *gin.Context) {
	raw := strings.TrimPrefix(c.Param("fileId"), "/")
	fileID := strings.TrimSuffix(raw, "/content")
	if fileID == "" {
		api.BadRequest(c, "fileId is required")
		return
	}
	var req struct {
		Content string `json:"content"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		api.BadRequest(c, err.Error())
		return
	}
	repo := repository.GetStorage()
	if err := repo.SaveFileContent(fileID, req.Content); err != nil {
		api.InternalError(c, err)
		return
	}
	api.Success(c, model.SaveFileContentResponse{Success: true, UpdatedAt: time.Now()})
}

// CreateFile POST /api/files/create
func CreateFile(c *gin.Context) {
	var req struct {
		ParentID string `json:"parentId"`
		Name     string `json:"name"`
		Content  string `json:"content"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		api.BadRequest(c, err.Error())
		return
	}
	if req.Name == "" {
		api.BadRequest(c, "name is required")
		return
	}
	repo := repository.GetStorage()
	newID, err := repo.CreateFile(req.ParentID, req.Name, req.Content)
	if err != nil {
		api.InternalError(c, err)
		return
	}
	api.Success(c, gin.H{"id": newID, "name": req.Name})
}

// CreateFolder POST /api/files/mkdir
func CreateFolder(c *gin.Context) {
	var req struct {
		ParentID string `json:"parentId"`
		Name     string `json:"name"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		api.BadRequest(c, err.Error())
		return
	}
	if req.Name == "" {
		api.BadRequest(c, "name is required")
		return
	}
	repo := repository.GetStorage()
	newID, err := repo.CreateFolder(req.ParentID, req.Name)
	if err != nil {
		api.InternalError(c, err)
		return
	}
	api.Success(c, gin.H{"id": newID, "name": req.Name})
}

// MoveNode POST /api/files/move
func MoveNode(c *gin.Context) {
	var req struct {
		ID          string `json:"id"`
		NewParentID string `json:"newParentId"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		api.BadRequest(c, err.Error())
		return
	}
	repo := repository.GetStorage()
	newID, err := repo.MoveNode(req.ID, req.NewParentID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	api.Success(c, gin.H{"id": newID})
}

// RenameNode POST /api/files/rename
func RenameNode(c *gin.Context) {
	var req struct {
		ID      string `json:"id"`
		NewName string `json:"newName"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		api.BadRequest(c, err.Error())
		return
	}
	repo := repository.GetStorage()
	newID, err := repo.RenameNode(req.ID, req.NewName)
	if err != nil {
		api.InternalError(c, err)
		return
	}
	api.Success(c, gin.H{"id": newID})
}

// DeleteNode DELETE /api/files/*fileId
func DeleteNode(c *gin.Context) {
	fileID := strings.TrimPrefix(c.Param("fileId"), "/")
	// 去掉可能的 /content 后缀（防止误删）
	fileID = strings.TrimSuffix(fileID, "/content")
	if fileID == "" {
		api.BadRequest(c, "fileId is required")
		return
	}
	repo := repository.GetStorage()
	if err := repo.DeleteNode(fileID); err != nil {
		api.InternalError(c, err)
		return
	}
	api.Success(c, gin.H{"success": true})
}

// SearchFiles GET /api/files/search?q=xxx
func SearchFiles(c *gin.Context) {
	query := c.Query("q")
	repo := repository.GetStorage()
	results := repo.SearchFiles(query)
	if results == nil {
		results = []repository.SearchResult{}
	}
	api.Success(c, results)
}

File: src/api/client/file.ts

import { ApiResponse, GetFileContentResponse, SaveFileContentResponse } from "./types";
import apiClient, { extractData, handleApiError } from "./utils";

/**
 * 获取单个文件的内容
 * fileId 是相对路径，如 "note.md" 或 "h1/note.md"
 */
export async function getFileContent(
  fileId: string,
): Promise<GetFileContentResponse> {
  try {
    const response = await apiClient.get<ApiResponse<GetFileContentResponse>>(
      `/file/${fileId}/content`,
    );
    return extractData(response);
  } catch (error) {
    handleApiError(error);
  }
}


/**
 * 搜索文件（文件名 + 内容）
 */
export async function searchFiles(query: string): Promise<Array<{ id: string; name: string; snippet: string; matchType: "name" | "content" }>> {
  try {
    const response = await apiClient.get<ApiResponse<Array<{ id: string; name: string; snippet: string; matchType: "name" | "content" }>>>(
      "/files/search",
      { params: { q: query } },
    );
    return extractData(response);
  } catch (error) {
    handleApiError(error);
  }
}


/**
 * 保存单个文件的内容
 */
export async function saveFileContent(
  fileId: string,
  content: string,
): Promise<SaveFileContentResponse> {
  try {
    const response = await apiClient.put<ApiResponse<SaveFileContentResponse>>(
      `/file/${fileId}/content`,
      { content },
    );
    return extractData(response);
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * 在服务端创建文件，返回真实 ID（相对路径）
 */
export async function createFileOnServer(
  parentId: string,
  name: string,
  content: string,
): Promise<{ id: string; name: string }> {
  try {
    const response = await apiClient.post<ApiResponse<{ id: string; name: string }>>(
      "/files/create",
      { parentId, name, content },
    );
    return extractData(response);
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * 在服务端创建文件夹，返回真实 ID（相对路径）
 */
export async function createFolderOnServer(
  parentId: string,
  name: string,
): Promise<{ id: string; name: string }> {
  try {
    const response = await apiClient.post<ApiResponse<{ id: string; name: string }>>(
      "/files/mkdir",
      { parentId, name },
    );
    return extractData(response);
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * 移动节点到新父目录，返回新 ID
 */
export async function moveNodeOnServer(
  id: string,
  newParentId: string,
): Promise<{ id: string }> {
  try {
    const response = await apiClient.post<ApiResponse<{ id: string }>>(
      "/files/move",
      { id, newParentId },
    );
    return extractData(response);
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * 重命名节点，返回新 ID
 */
export async function renameNodeOnServer(
  id: string,
  newName: string,
): Promise<{ id: string }> {
  try {
    const response = await apiClient.post<ApiResponse<{ id: string }>>(
      "/files/rename",
      { id, newName },
    );
    return extractData(response);
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * 删除节点（文件或目录）
 */
export async function deleteNodeOnServer(id: string): Promise<void> {
  try {
    await apiClient.delete(`/file/${id}`);
  } catch (error) {
    handleApiError(error);
  }
}
